### PR TITLE
Fix query binding with two colons in query

### LINKED ISSUE #5114: Bug: Query binding fails when date_format is set with '%H:%i:%s'
**Describe the bug**
Query builder is binding incorrectly the statement with date_format (eg. %d/%m/%Y %H:%i:%s) - 22/09/2021 10:45:00 and crash.

```php
<?php

$id = 1;
//Works
$sql = " SELECT mytable.id, DATE_FORMAT(mytable.created_at,'%d/%m/%Y %H:%i') AS created_at_uk FROM mytable WHERE mytable.id = ?";
$query = $this->db->query($sql, [$id]);

//DOES NOT work
$sql = " SELECT mytable.id, DATE_FORMAT(mytable.created_at,'%d/%m/%Y %H:%i:%s') AS created_at_uk FROM mytable WHERE mytable.id = ?";
$query = $this->db->query($sql, [$id]);

```
As you can see, the only difference between both SQL is the `:%s`. This should not be binding as the (%s) means seconds on date_format.

**CodeIgniter 4 version**
CI 4.1.3

**Affected module(s)**
Not sure but looks like is Query.php

**Expected behavior, and steps to reproduce if appropriate**
Should return the query with day/month/year hour:minute:second

**Context**
 - OS: Mac
 - Web server Apache
 - PHP version 8.0

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
Fixes #5114

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Review by MGatner — APPROVED** [review]
This looks good to me but Database is my "blind spot" so I'd like @paulbalandan or @samsonasik to take a look.

File: system/Database/Query.php [hunk 1]
```diff
@@ -275,11 +275,13 @@ protected function compileBinds()
     {
         $sql = $this->finalQueryString;
 
-        $hasNamedBinds = preg_match('/:((?!=).+):/', $sql) === 1;
+        $hasBinds      = strpos($sql, $this->bindMarker) !== false;
+        $hasNamedBinds = ! $hasBinds
+            && preg_match('/:(?!=).+:/', $sql) === 1;
```
Comment (samsonasik) [inline system/Database/Query.php:280]: Could you add `@see https://regex101.com` link for the regex update, it  can be reference in the future

File: system/Database/Query.php [hunk 2]
```diff
@@ -275,11 +275,13 @@ protected function compileBinds()
...
 
         if (empty($this->binds)
             || empty($this->bindMarker)
-            || (! $hasNamedBinds && strpos($sql, $this->bindMarker) === false)
+            || (! $hasNamedBinds && ! $hasBinds)
```
Comment (paulbalandan) [inline system/Database/Query.php:284]: $hasBinds is essentially duplicated here. Instead, we can either:
1. Interchange the operands, with `strpos` being called first
2. Retain $hasNamedBinds only, as it contains already the $hasBinds check

**Review by paulbalandan — CHANGES_REQUESTED** [review]

File: system/Database/Query.php (hunk 2 above)
Comment (kenjis) [reply system/Database/Query.php:284]: @paulbalandan 
1
I don't understand your intention. I removed duplicated `strpos()` call.

2
If I remove `&& ! $hasBinds`, it breaks.
https://3v4l.org/UWTvF

File: system/Database/Query.php (hunk 1 above)
Comment (kenjis) [reply system/Database/Query.php:280]: @samsonasik I updated the link.

File: system/Database/Query.php (hunk 2 above)
Comment (paulbalandan) [reply system/Database/Query.php:284]: Okay. I stand corrected with my second suggestion. Basically, I want something like this.
https://3v4l.org/V4pJv

**Review by paulbalandan — APPROVED** [review]
I overlooked a detail in the later part. Sorry for the noise.

**Comment by paulbalandan** [comment]
Thank you, @kenjis

**Comment by vlakoff** [comment]
I have checked the differences between the two codes, and I don't see where is the functional difference.

The change in code is that a `! $hasBinds` condition has been added when determining `$hasNamedBinds`,
but nevertheless the early return is triggered when both `! $hasNamedBinds && ! $hasBinds`.

So, if `$hasBinds` is true, there can't be early return, and if `$hasBinds` is false, `$hasNamedBinds` gets determined as previously.

**Comment by kenjis** [comment]
What are the two codes?

`$hasBinds` means the query contains bindings, that is `?`.
`$hasNamedBinds` means the query contains named bindings, that is something like`:name:`.
They are exclusive. We cannot use both at a time.
So I added `! $hasBinds` condition when determining `$hasNamedBinds`.

`! $hasNamedBinds && ! $hasBinds` means the query does not have `:name:` nor `?`.
That is no bindings at all. So we have nothing to process bindings.
So we go early return.

**Comment by vlakoff** [comment]
I understood, what your change influences is actually a line that is a bit below in the code:

```php
$sql = $hasNamedBinds ? $this->matchNamedBinds($sql, $binds) : $this->matchSimpleBinds($sql, $binds, $bindCount, $ml);
```

**Comment by vlakoff** [comment]
Though, the code would fail with named binds, simply if there is a "?" somewhere in the supplied SQL string.

As a quick example:
```sql
SELECT * FROM songs WHERE band = :band: AND title = 'Do You Love Me?'
```

`hasBinds` would be erroneously true, which in turn would make `hasNamedBinds` erroneously false.

**Comment by vlakoff** [comment]
Refs #5138.

**Pull request merged**